heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/metricscache/MetricsCacheQueryUtils.java

```diff
@@ -0,0 +1,339 @@
+            intervalValueBuilder.setInterval(TopologyMaster.MetricInterval.newBuilder()
+                .setStart(v.startTime).setEnd(v.endTime));
+
+            individualMetricBuilder.addIntervalValues(intervalValueBuilder);
+          }// end IntervalValue
+        }
+
+        taskMetricBuilder.addMetric(individualMetricBuilder);
+      }// end IndividualMetric
+
+      builder.addMetric(taskMetricBuilder);
+    }// end TaskMetric
+
+    return builder.build();
+  }
+
+  // compatible with com.twitter.heron.proto.tmaster.TopologyMaster.ExceptionLogRequest
+  public static ExceptionRequest fromProtobuf(TopologyMaster.ExceptionLogRequest request) {
+    String componentName = request.getComponentName();
+
+    ExceptionRequest outRequest = new ExceptionRequest();
+    outRequest.componentNameInstanceId = new HashMap<>();
+    Set<String> instances = null;
+
+    if (request.getInstancesCount() > 0) {
+      instances = new HashSet<>();
+      instances.addAll(request.getInstancesList());
+    } // empty list means all instances; 'null' means all instances
+
+    // only one component
+    outRequest.componentNameInstanceId.put(componentName, instances);
+
+    return outRequest;
+  }
+
+  // compatible with com.twitter.heron.proto.tmaster.TopologyMaster.ExceptionLogResponse
+  public static TopologyMaster.ExceptionLogResponse toProtobuf(ExceptionResponse response) {
+    TopologyMaster.ExceptionLogResponse.Builder builder =
+        TopologyMaster.ExceptionLogResponse.newBuilder();
+    // default OK if we have response to build already
+    builder.setStatus(Common.Status.newBuilder().setStatus(Common.StatusCode.OK));
+
+    for (ExceptionDatapoint e : response.exceptionDatapointList) {
+      TopologyMaster.TmasterExceptionLog.Builder exceptionBuilder =
+          TopologyMaster.TmasterExceptionLog.newBuilder();
+      // ExceptionDatapoint
+      exceptionBuilder.setComponentName(e.componentName);
+      exceptionBuilder.setHostname(e.hostname);
+      exceptionBuilder.setInstanceId(e.instanceId);
+      // ExceptionData
+      exceptionBuilder.setStacktrace(e.stackTrace);
+      exceptionBuilder.setLasttime(e.lastTime);
+      exceptionBuilder.setFirsttime(e.firstTime);
+      exceptionBuilder.setCount(e.count);
+      exceptionBuilder.setLogging(e.logging);
+
+      builder.addExceptions(exceptionBuilder);
+    }
+
+    return builder.build();
+  }
+
+
+  public enum Granularity {
+    AGGREGATE_ALL_METRICS,
+    AGGREGATE_BY_BUCKET,
+    RAW
+  }
+
+  public static class ExceptionRequest {
+    public Map<String, Set<String>> componentNameInstanceId;
+  }
+
+  public static class ExceptionResponse {
+    public List<ExceptionDatapoint> exceptionDatapointList;
+  }
+
+  /**
+   * immutable data bag for time range value
+   * time window: startTime ~ endTime, in milli-seconds
+   * metric value string: value
+   */
+  public static final class MetricTimeRangeValue {
+    public final long startTime;
```

Comment:
Sorry, I missed the final part. Yes, this would be immutable, but the idiomatic java way is to make the members private, pass them all in via a constructor and access them via getter methods. That makes the immutable contract clear when looking at the constructor and method declarations. In that way the contract is clearer when studying the code or the javadocs. It also lends itself to future refactoring where subclassing or decoration is required.